### PR TITLE
Fix Issue 16744: Add Typeof (Redux)

### PR DESCRIPTION
**Pull request by MetaLang**
Previous PR: https://github.com/dlang/phobos/pull/5662. I'm resubmitting this after it was reverted in https://github.com/dlang/phobos/pull/5664.

This an alternative implementation to #4920.

### CODE REVIEW

File: std/traits.d [hunk 1]
```diff
@@ -5548,6 +5549,132 @@ template BuiltinTypeOf(T)
...
+    else static if (is(T[0] TypeOfT0))
+        alias Typeof = TypeOfT0;
+    else static if (is(typeof(T[0]) TypeOfT0))
+        alias Typeof = TypeOfT0;
+    else
+        static assert(false, "Cannot take the type of " ~ T[0].stringof);
+}
+
+///
+@safe unittest
```
Comment (atilaneves) [inline std/traits.d]: I'd add `pure` and convert the `assert`s into `static assert`s.

Comment (MetaLang) [reply std/traits.d]: I don't see a point in adding pure as `Typeof` is a template, not a function.

Do we have a policy on using `static assert` vs. `assert` in unit tests for stuff that can be checked at compile time? I don't know if it's necessarily a good idea for _compilation_ to fail when building with tests enabled, but that's just a vague sense of unease on my part.

Comment (andralex) [reply std/traits.d]: static assert is better than assert but keep in in the unittest so it doesn't slow down regular compilation

**Comment by andralex** [comment]
Wished there were more examples where `Typeof` saves the day. Are there examples without `staticMap`? If not, perhaps `mapTypes` would be a better way to go.

File: std/traits.d (hunk 1 above)
Comment (MetaLang) [reply std/traits.d]: I've changed all asserts to static asserts.

**Comment by MetaLang** [comment]
@andralex I've run into this a lot in template-heavy generic code. I'll try and find a few examples.

This could easily be done in user code as it's a very small template to define, but IMO the value in having this in the standard library is that it handles non-obvious edge cases, such as trying to pass a module or `__error__`.

Now that I think about it, however, this also reifies the `typeof(template) -> void` behaviour of DMD, which may(?) change in the future. Should I remove that and just return whatever regular `typeof` returns, or should this behaviour be kept?

**Comment by andralex** [comment]
I'd want to focus on something higher-level and useful instead of something that is just like typeof but a bit different. Those fare poorly because people are never sure which one they're supposed to use etc. In contrast, a higher-level primitive will shine by its own utility.

That's why I'm saying: if most of the time the use of `Typeof` is in conjunction with `staticMap`, we're better off with a template `Types`. When given a list of types, it yields an `AliasSeq` containing the type of each. No type? The empty `AliasSeq`. One type? AliasSeq with one element. Perhaps that would have more use cases.

**Comment by MetaLang** [comment]
@andralex you're confusing me now, because you seem to be talking about something different from what's requested in the bug report. `Typeof` does the exact same thing as `typeof`, except when passed a type it just aliases itself to that type instead of causing a compile error (if you try to do `typeof(int)`, it won't compile), and it can be passed as a normal symbol (i.e., `staticMap!(Typeof, <list of values and/or types>)` will work whereas `staticMap!(typeof, <list of values and/or types>)` will not).

To my knowledge, `Typeof` has the same behaviour as `typeof` in every other respect. If you pass a symbol that's a module name to `Typeof`, or the special `__error__` symbol, it will generate a compile-time error just like `typeof` would. Likewise, if you pass a template to `Typeof`, it'll return `void` just like `typeof` currently does (the only reason I brought it up is because I'm pretty sure this is an implementation detail and not in the spec).

**Comment by andralex** [comment]
> @andralex you're confusing me now, because you seem to be talking about something different from what's requested in the bug report.

Yes. More specifically: I am trying to resolutely steer away from "just about the same, only a bit different".

**Comment by MetaLang** [comment]
Okay, then the enhancement request should be closed as WONTFIX. I'll close this PR and close the bug.

**Pull request closed without merging**